File: CK3toEU4/Source/CK3World/CoatsOfArms/CoatsOfArms.cpp
#include "CoatsOfArms.h"
#include "../Titles/Title.h"
#include "../Titles/Titles.h"
#include "CoatOfArms.h"
#include "Log.h"
#include "ParserHelpers.h"

CK3::CoatsOfArms::CoatsOfArms(std::istream& theStream)
{
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}

void CK3::CoatsOfArms::registerKeys()
{
	registerRegex(R"(\d+)", [this](const std::string& coaID, std::istream& theStream) {
		auto newCoA = std::make_shared<CoatOfArms>(theStream, std::stoll(coaID));
		coats.insert(std::pair(newCoA->getID(), newCoA));
	});
	registerKeyword("coat_of_arms_manager_database", [this](const std::string& unused, std::istream& theStream) {
		coats = CoatsOfArms(theStream).getCoats();
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}

void CK3::CoatsOfArms::linkParents(const Titles& titles)
{
	auto counter = 0;
	const auto& titleData = titles.getTitles();
	for (const auto& coat: coats)
	{
		if (coat.second->getParent().first.empty())
			continue;
		const auto& titleDataItr = titleData.find(coat.second->getParent().first);
		if (titleDataItr != titleData.end())
		{
			if (!titleDataItr->second->getCoA())
				throw std::runtime_error("CoA " + std::to_string(coat.first) + " has parent " + coat.second->getParent().first + " which has no coat defined!");
			if (!coats.count(titleDataItr->second->getCoA()->first))
				throw std::runtime_error("CoA " + std::to_string(coat.first) + " has parent " + coat.second->getParent().first + " which has invalid coat defined!");
			coat.second->loadParent(std::make_pair(coat.second->getParent().first, coats[titleDataItr->second->getCoA()->first]));
			++counter;
		}
		else
		{
			throw std::runtime_error("CoA " + std::to_string(coat.first) + " has parent " + coat.second->getParent().first + " which is undefined!");
		}
	}
	Log(LogLevel::Info) << "<> " << counter << " coats updated.";
}

File: CK3ToEU4/Source/CK3/CoatsOfArms/CoatsOfArms.cs

using System;
using System.Collections.Generic;
using commonItems;
using commonItems.Colors;

namespace CK3ToEU4.CK3.CoatsOfArms;

class CoatsOfArms
{
	public CoatsOfArms()
	{
	}

	public CoatsOfArms(BufferedReader reader, ColorFactory colorFactory)
	{
		var parser = new Parser();
		registerKeys(parser, colorFactory);
		parser.ParseStream(reader);
	}
	public IReadOnlyDictionary<long, CoatOfArms> Coats => coats;
	private Dictionary<long, CoatOfArms> coats;

	public void linkParents(Titles.Titles titles)
	{
		int counter = 0;
		var titleData = titles.getTitles();
		foreach (var coat in coats)
		{
			if (coat.Value.Parent is null)
				continue;
			const auto& titleDataItr = titleData.find(coat.second->getParent()->first);
			if (titleDataItr != titleData.end())
			{
				if (!titleDataItr->second->getCoA())
					throw new Exception("CoA " + std::to_string(coat.first) + " has parent " + coat.second->getParent()->first + " which has no coat defined!");
				if (!coats.count(titleDataItr->second->getCoA()->first))
					throw new Exception(
						"CoA " + std::to_string(coat.first) + " has parent " + coat.second->getParent()->first + " which has invalid coat defined!");
				coat.second->loadParent(std::make_pair(coat.second->getParent()->first, coats[titleDataItr->second->getCoA()->first]));
				++counter;
			}
			else
			{
				throw new Exception("CoA " + std::to_string(coat.first) + " has parent " + coat.second->getParent()->first + " which is undefined!");
			}
		}
		Logger.Info("<> " + counter + " coats updated.");
	}

	:
	private void registerKeys(Parser parser, ColorFactory colorFactory)
	{
		parser.RegisterRegex(CommonRegexes.Integer, (reader, coaID) => {
			var newCoA = new CoatOfArms(reader, long.Parse(coaID), colorFactory);
			coats[newCoA.Id] = newCoA;
		});
		parser.RegisterKeyword("coat_of_arms_manager_database", reader => {
			coats = new (new CoatsOfArms(reader, colorFactory).Coats);
		});
		parser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreItem);
	}

};